src/storage.py

```diff
@@ -0,0 +1,283 @@
+        """Checks if the subprocess error is caused by safety check."""
+        return "need at least 3 OSDs" in err
+
+    def _run(self, cmd: list) -> str:
+        """Wrapper around subprocess run for storage commands."""
+        process = run(cmd, capture_output=True, text=True, check=True, timeout=180)
+        logger.debug(f"Command {' '.join(cmd)} finished; Output: {process.stdout}")
+        return process.stdout
+
+    def _enroll_with_db(self, disk: list, db: list):
+        """Checks if sufficient devices are available to be enrolled into OSDs."""
+        enrollment_count = min(len(disk), len(db))
+        for i in range(enrollment_count):
+            try:
+                microceph.add_osd_cmd(spec=disk[i], db_dev=db[i])
+                # store configured devices.
+                self._save_osd_data(disk=disk[i], db=db[i])
+            except CalledProcessError as e:
+                logger.error(e.stderr)
+
+    def _enroll_disks_in_batch(self, disks: list):
+        """Adds requested Disks to Microceph and stored state."""
+        microceph.enroll_disks_as_osds(disks)
+        for disk in disks:
+            self._save_osd_data(disk)
+        logger.debug(f"Added {disks} as OSDs.")
+
+    def remove_osd(self, osd: int, force: bool = False):
+        """Removes OSD from MicroCeph and from stored state."""
+        try:
+            microceph.remove_disk_cmd(osd, force)
+            self._remove_osd_data(self._stored.osd_data[osd])
+        except CalledProcessError as e:
+            if force:
+                # clean stored state.
+                self._remove_osd_data(self._stored.osd_data[osd])
+            raise e
+
+    def _save_osd_data(self, disk: str, db: str = None):
+        """Save OSD data."""
+        for osd in microceph.list_disk_cmd()["ConfiguredDisks"]:
+            # get block device info using /dev/disk-by-id and lsblk.
+            local_device = microceph._get_disk_info(osd["path"])
+
+            # OSD not configured on current unit.
+            if not local_device:
+                continue
+
+            # e.g. check 'vdd' in '/dev/vdd'
+            if local_device["name"] in disk:
+                self._stored.osd_data[osd["osd"]] = {
+                    "disk_by_id": osd["path"],
+                    "disk": disk,
+                    "db": db,
+                }
+
+    def _get_osd_num(self, disk, directive):
+        """Fetch the OSD number of consuming OSD, None is not used as OSD."""
+        # both standalone and dedicated db disks are used as OSD disks.
+        if directive in [self.standalone, self.dedicated_db]:
+            directive = "disk"
+
+        logger.debug(self._stored.osd_data)
+        logger.debug(f"Incoming Disk {disk}, directive {directive}.")
+
+        for k, v in dict(self._stored.osd_data).items():
+            if v and v[directive] == disk:
+                return k  # key is the stored osd number.
+        return None
+
+    def _clean_stale_osd_data(self):
```

Comment:
Afaict this doesn't account for DB devices, right?

Reply:
This functions runs a pass through configured OSDs and removes any OSD entry (which may have DB key) that is not present in MicroCeph records. Removing a configured DB devices WILL result in OSD removal so if such an entry was not cleaned up (fsr) it will clean that up too.